syntax = "proto3";

package brain.state.v1;

import "brain/shared/v1/envelope.proto";
import "google/protobuf/timestamp.proto";

// Authority service for durable blob/object storage.
service ObjectAuthorityService {
  rpc PutObject(PutObjectRequest) returns (PutObjectResponse);
  rpc GetObject(GetObjectRequest) returns (GetObjectResponse);
  rpc StatObject(StatObjectRequest) returns (StatObjectResponse);
  rpc DeleteObject(DeleteObjectRequest) returns (DeleteObjectResponse);
}

// Canonical content-addressed object key reference.
message ObjectRef {
  string object_key = 1;
}

// Materialized object metadata.
message ObjectRecord {
  ObjectRef ref                         = 1;
  uint64 size_bytes                     = 2;
  string digest_algorithm               = 3;
  string digest_hex                     = 4;
  string content_type                   = 5;
  google.protobuf.Timestamp created_at  = 6;
}

message PutObjectRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  bytes content                        = 2;
  string content_type                  = 3;
}

message PutObjectResponse {
  brain.shared.v1.EnvelopeMeta meta           = 1;
  ObjectRecord object                           = 2;
  bool already_exists                           = 3;
  repeated brain.shared.v1.ErrorDetail errors = 4;
}

message GetObjectRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  ObjectRef ref                        = 2;
}

message GetObjectResponse {
  brain.shared.v1.EnvelopeMeta meta           = 1;
  ObjectRecord object                           = 2;
  bytes content                                 = 3;
  repeated brain.shared.v1.ErrorDetail errors = 4;
}

message StatObjectRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  ObjectRef ref                        = 2;
}

message StatObjectResponse {
  brain.shared.v1.EnvelopeMeta meta           = 1;
  bool found                                    = 2;
  ObjectRecord object                           = 3;
  repeated brain.shared.v1.ErrorDetail errors = 4;
}

message DeleteObjectRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  ObjectRef ref                        = 2;
  bool missing_ok                      = 3;
}

message DeleteObjectResponse {
  brain.shared.v1.EnvelopeMeta meta           = 1;
  bool deleted                                  = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}
